### DISCUSSION TITLE
CI speedup overview

### DISCUSSION THREAD

**Discussion by pjonsson** [picolibc/picolibc#572] [Ideas]
# Overview

Speeding up the CI consists of some (mostly) orthogonal lines of work:

- #553, #559 (reduce load times in each CI run, free up space for other things in the Github cache area)
- #558 (Peter measured do-arm-configure and got 3.8x speedup with warm cache)
- Consolidate/re-organize job composition to avoid load times/improve ccache hit rates (partly fixed with #555 and #562)
- Add ccache to the Docker images & Linux jobs (Peter has patches)
- Cancel previous iteration of running jobs when a PR is updated

# Technical notes

## CI Runners

The documented limit for concurrent runners for the free level is 20, where at most 5 can be macOS.

## Cache

The Github cache is 10 GB and evicts with LRU policy. With the reduced image sizes in #553, there is 7.1 GB of space available for ccache.

The do-arm-configure measurement filled ccache to 7.95% of 5 GB (just under 400 MB). Assuming all other configurations have the same size, that gives room for 17 distinct ccaches before Github starts evicting the working set.

The cache can be compressed a bit (~150 MB -> 95 MB for ARM object files).

# Miscellaneous Ideas/Questions

Commit d2c4079f5f440 mentions a 2 GB limit, but looking at the current [cache](https://github.com/picolibc/picolibc/actions/caches), dockerimg-zephyr.tar.xz on main is 2.1GB, and the old dockerimg.tar.xz is 2.3GB. Building a single image of everything is 8.83GB, which (assuming same level of compression as for the Zephyr image) compresses to 2.8GB.

If file size in the cache is an issue, can split be used to produce multiple files that are stored in their separate cache-key?

Can the Github [container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) eliminate the need for the images to be in the 10 Gb cache?

## Zstd compression levels

Compress a combined linux and zephyr image.

Conclusion 1: Compression level should not be chosen based on load times. Before ultra levels, decompression time adds 2-3 seconds of time to docker load. Even at ultra level, the extra time is 12s. 

Conclusion 2: `zstd -T2 -18` is significantly cheaper than `zstd -T20 -18` (2000 CPU seconds vs 3300 CPU seconds).

Compression level 17 brings the size down below 2GB.

Increasing compression level from 17 to 18 costs 600 CPU seconds, and saves 80MB.

Increasing compression level from 18 to 19 costs 1100 CPU seconds, and saves 40MB. Questionable price/performance ratio.

Ultra levels: increasing compression level from 19 to 20 costs 200 CPU seconds, and saves 500MB. Comparing the more reasonable level 18 to 20 costs 1300 CPU seconds, and saves 540MB. Is 540MB additional space savings worth 650 seconds?

Ultra levels 21/22 ran out of memory on my machine.

```
$ time -p docker load -i file.tar
Loaded image: picolibc-full:latest
real 12,81
user 0,82
sys 5,87
$ time -p zstd -T0 -14 file.tar
file.tar             : 24.87%   (8899393024 => 2213562358 bytes, file.tar.zst)
real 81,44
user 1265,21
sys 25,43
$ time -p docker load -i file.tar.zst
Loaded image: picolibc-full:latest
real 13,19
user 0,24
sys 2,35
$ time -p zstd -T0 -17 file.tar
file.tar             : 22.34%   (8899393024 => 1988474044 bytes, file.tar.zst)
real 172,77
user 2763,71
sys 49,49
$ time -p docker load -i file.tar.zst
Loaded image: picolibc-full:latest
real 14,88
user 0,22
sys 2,14
$ time -p zstd -T0 -18 file.tar
file.tar             : 21.40%   (8899393024 => 1904806171 bytes, file.tar.zst)
real 198,58
user 3311,20
sys 61,09
$ time -p docker load -i file.tar.zst
Loaded image: picolibc-full:latest
real 14,81
user 0,19
sys 2,03
$ time -p zstd -T0 -19 file.tar 
file.tar             : 21.20%   (8899393024 => 1886366685 bytes, file.tar.zst) 
real 253,80
user 4450,65
sys 84,55
$ time -p docker load -i file.tar.zst 
Loaded image: picolibc-full:latest
real 14,85
user 0,19
sys 2,06
$ time -p zstd -T0 --ultra -20 file.tar
file.tar             : 15.27%   (8899393024 => 1358600452 bytes, file.tar.zst)
real 288,94
user 4613,78
sys 77,09
$ time -p docker load -i file.tar.zst
Loaded image: picolibc-full:latest
real 24,36
user 0,11
sys 1,27
```

**Comment by keith-packard** [comment]
>     * Consolidate/re-organize job composition to avoid load times/improve ccache hit rates

Yeah, this seems like a pretty straightforward plan -- for all of the "small" targets, do all of the build variations in a single run. We would enable ccache, but not bother saving it. Mix as many small targets into a single build as will run in a reasonable amount of time. For the three large targets (arm, risc-v and (now) m68k), create a persistent ccache for each and then see how many variations we can run in a similar amount of time. I'd bet with ccache we will be able to run 2-3 in under an hour.

I think we should also change the build style selection from fortify+minsize, minsize, release to minsize, fortify+release. And, we can blend more of the variants together as their code doesn't overlap -- the math library configurations do not interact with the stdio configurations. I think we can get down to 4-5 build option varations from 10.

> The Github cache is 10 GB and evicts with LRU policy. With the reduced image sizes in #553, there is 7.1 GB of space available for ccache.

I did a run of all arm variants (all 10 current build flag settings), the cache is 1.8GB at the end. If we only cache arm/risc-v and m68k across builds, we should have enough space.

> Commit [d2c4079](https://github.com/picolibc/picolibc/commit/d2c4079f5f4405fedde4ff5c80503c99a9b79da0) mentions a 2 GB limit

I think that was relaxed in cache@v3. I was seeing failures before switching versions.

> Conclusion 1: Compression level should not be chosen based on load times. Before ultra levels, decompression time adds 2-3 seconds of time to docker load. Even at ultra level, the extra time is 12s.

Given that the caches can last for weeks, I think we can allow the compression time to grow considerably. I think the real limit is the 10GB total cache space. Adjust this value to keep under the total budget while balancing the space with ccache.

Thanks for doing the compression analysis, that gives us good data for when we start bumping up against the total space available.

**Comment by pjonsson** [comment]
If the limitations of cache@v2 have been lifted with the current version, are you against merging the two Dockerfiles again (later)? (My compression tests were on a merged Dockerfile, so the full docker image can be 1.35GB with level 20 zstd compression.) It will be terrible to wait for the compression, but if that is done once a month, it is probably a good tradeoff as you say. Oh, and the current images in the cache are called xz, but I'm fairly sure they are uncompressed tar files with an xz extension.

The 1.8GB cache you measured should compress some, so with a bit of luck, it will be 1.3GB in the cache.

I'll add ccache to the CI after you have merged #555.

**Comment by keith-packard** [comment]
No, merging them together would be fine if you think the savings are worth it. I was thinking the lack of any overlap between toolchains would make running each instance enough faster to be worth splitting though. We do occasionally get timeouts caused by the large cache file download time...

**Comment by pjonsson** [comment]
When I wrote my question, my perspective was mostly about simplicity since I believe eliminating decision points is by far the best way to encourage hesitant observers to make the leap into becoming occasional contributors. A single image makes it easy to explain how to run tests, and where to add additional software required to run new tests.

I'm starting to wonder if the single image is more or less a requirement for slicing the CI jobs to be on a per architecture split instead.

Speaking of images, are `.github/debian-packages.txt` and `.github/packages.txt` used?

**Comment by pjonsson** [comment]
What do you think about cancelling the previous jobs of a PR when pushing an update?

I haven't tested, but combining https://docs.github.com/en/actions/using-jobs/using-concurrency#example-only-cancel-in-progress-jobs-or-runs-for-the-current-workflow and the `github.head_ref || github.run_id` fallback value described at https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-a-fallback-value looks promising.

**Comment by keith-packard** [comment]
> What do you think about cancelling the previous jobs of a PR when pushing an update?

That would be awesome. I just don't know how to make it work.

**Comment by pjonsson** [comment]
I updated the bullet list at the start, and it seems like we've made quite a lot of progress.

The remaining major thing is fitting the cache for the Linux&Zephyr jobs in the 10GB limit. You described one possibility of re-organizing the jobs on a per-architecture basis instead but that it might result in some quite short and some really long jobs. I think that strategy would enable the most cache sharing, and if there is sufficient space in the Github cache, the longest jobs could be split into 2 or 3 jobs.

We currently have >8GB "free space" in the cache, so an individual cache for each of the zephyr jobs can easily fit in in that. It would be better if the Linux jobs could be cached, but that doesn't fit. Since the number of concurrent runners is also an issue, I will make a PR which speeds up the Zephyr jobs as a start.

I'll take a look at cancelling the jobs when a PR is updated, too.
